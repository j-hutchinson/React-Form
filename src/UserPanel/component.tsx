import React, { FormEvent, useState } from 'react';
import styled from 'styled-components';
import { PanelStyles, SubmitInputStyles } from '../css/tokens';
import { updateUser } from '../store/actions/actions';
import store from '../store/store';

interface Props {
    nextPanel: () => void;
}

export const StyledPanel = styled.form`
    ${PanelStyles}
`;

const StyledLabel = styled.label`
    display: grid;
    grid-template-rows: min-content 1fr;
    grid-gap: 4px;
    margin: 10px;
`;

const StyledButtonWrapper = styled.div`
    display: flex;
    flex-direction: row-reverse;
    margin-right: 10px;
`;

const StyledAsterisk = styled.div`
    display: inline-block;
    color: red;
`;

const StyledError = styled.div`
    color: red;
    font-size: 12px;
    margin: 10px;
`;

const StyledSubmitInput = styled.input`
   ${SubmitInputStyles};
`;

export const StyledInput = styled.input`
   height: 24px;
   font-size: 14px;
`;

/*
**  Check if password contains a digit, a lower case letter, an upper case letter, and greater than or equal to 9 characters in length
*/
export const passwordValidation = (password: string): boolean => {
    return /\d/.test(password) && /[a-z]/.test(password) && /[A-Z]/.test(password) && password.length >= 9
}

const UserPanel = ({ nextPanel }: Props): JSX.Element => {
    const [name, setName] = useState<string>('');
    const [role, setRole] = useState<string>('');
    const [email, setEmail] = useState<string>('');
    const [password, setPassword] = useState<string>('');
    const [error, setError] = useState<boolean>(false);

    const onSubmission = (ev: FormEvent): void => {
        ev.preventDefault();
        if (passwordValidation(password)) {
            store.dispatch(updateUser({ name, role, email, password }));
            nextPanel();
        } else {
            setError(true);
        }
    }

    return (
        <StyledPanel onSubmit={onSubmission}>
            <StyledLabel>
                <div> Name: <StyledAsterisk>*</StyledAsterisk> </div>
                <StyledInput type="text" value={name} onChange={(ev) => setName(ev.target.value)} required />
            </StyledLabel>
            <StyledLabel>
                Role:
                <StyledInput type="text" value={role} onChange={(ev) => setRole(ev.target.value)} />
            </StyledLabel>
            <StyledLabel>
                <div> Email: <StyledAsterisk>*</StyledAsterisk> </div>
                <StyledInput type="email" value={email} onChange={(ev) => setEmail(ev.target.value)} required />
            </StyledLabel>
            <StyledLabel>
                <div> Password: <StyledAsterisk>*</StyledAsterisk> </div>
                <StyledInput type="password" value={password} onChange={(ev) => setPassword(ev.target.value)} required />
            </StyledLabel>
            {error && (
                <StyledError>Check your password is valid. Must contain 1 uppercase, 1 lowercase, 1 number and be longer than 9 characters</StyledError>
            )}
            <StyledButtonWrapper>
                <StyledSubmitInput type="submit" value="Submit" />
            </StyledButtonWrapper>
        </StyledPanel>
    )
};

export default UserPanel;