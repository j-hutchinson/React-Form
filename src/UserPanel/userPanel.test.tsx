import { shallow } from 'enzyme';
import React from 'react';
import UserPanel, { StyledInput, StyledPanel, passwordValidation } from './component';
import store from '../store/store'

jest.mock('../store/store', () => ({ dispatch: jest.fn(), }));

describe('UserPanel component', () => {
    const preventDefault = jest.fn();

    afterEach(() => {
        jest.resetAllMocks();
    });

    test('component matches snapshot', () => {
        expect.assertions(1);

        const wrapper = shallow(<UserPanel />);

        expect(wrapper).toMatchSnapshot();
    });

    test('onSubmit updates the store with valid User object', () => {
        expect.assertions(3);

        const wrapper = shallow(<UserPanel />);

        wrapper.find(StyledInput).at(0).simulate('change', { target: { value: 'Jack Hutchinson' } });
        wrapper.find(StyledInput).at(1).simulate('change', { target: { value: 'Coder' } });
        wrapper.find(StyledInput).at(2).simulate('change', { target: { value: 'email@email.com' } });
        wrapper.find(StyledInput).at(3).simulate('change', { target: { value: 'ValidPassword123' } });
        wrapper.find(StyledPanel).props().onSubmit({ preventDefault });

        expect(store.dispatch).toHaveBeenCalledTimes(2);
        expect(store.dispatch).toHaveBeenCalledWith({
            user: {
                name: 'Jack Hutchinson',
                role: 'Coder',
                email: 'email@email.com',
                password: 'ValidPassword123'
            },
            type: "UPDATE_USER"
        });
        expect(store.dispatch).toHaveBeenCalledWith({ type: "NEXT_PAGE" });
    });

    test('Store is updated with missing role', () => {
        expect.assertions(3);

        const wrapper = shallow(<UserPanel />);

        wrapper.find(StyledInput).at(0).simulate('change', { target: { value: 'Jack Hutchinson' } });
        wrapper.find(StyledInput).at(1).simulate('change', { target: { value: null } });
        wrapper.find(StyledInput).at(2).simulate('change', { target: { value: 'email@email.com' } });
        wrapper.find(StyledInput).at(3).simulate('change', { target: { value: 'ValidPassword123' } });
        wrapper.find(StyledPanel).props().onSubmit({ preventDefault });

        expect(store.dispatch).toHaveBeenCalledTimes(2);
        expect(store.dispatch).toHaveBeenCalledWith({
            user: {
                name: 'Jack Hutchinson',
                role: null,
                email: 'email@email.com',
                password: 'ValidPassword123'
            },
            type: "UPDATE_USER"
        });
        expect(store.dispatch).toHaveBeenCalledWith({ type: "NEXT_PAGE" });
    });

    test('Store is not updated when password is invalid', () => {
        expect.assertions(2);

        const wrapper = shallow(<UserPanel />);

        wrapper.find(StyledInput).at(0).simulate('change', { target: { value: 'Jack Hutchinson' } });
        wrapper.find(StyledInput).at(1).simulate('change', { target: { value: 'Coder' } });
        wrapper.find(StyledInput).at(2).simulate('change', { target: { value: 'email@email.com' } });
        wrapper.find(StyledInput).at(3).simulate('change', { target: { value: 'not_valid' } });
        wrapper.find(StyledPanel).props().onSubmit({ preventDefault });

        expect(store.dispatch).toHaveBeenCalledTimes(0);
        expect(preventDefault).toHaveBeenCalledTimes(1);
    });
});

describe('Password validation checker', () => {
    test.each`
        password                                          | valid 
        ${'Password1'}                                    | ${true}        
        ${'password'}                                     | ${false}        
        ${'password1'}                                    | ${false}        
        ${'123456789'}                                    | ${false}        
        ${'Pass1'}                                        | ${false}        
        ${'PASSWORD'}                                     | ${false}        
        ${'PASSWORD123'}                                  | ${false}        
        ${'SuperComplicatedLongPasswordWithNoNumbers'}    | ${false}        
    `('passwordValidation for $password', ({ password, valid }) => {
        expect(passwordValidation(password)).toEqual(valid);
    });
});
