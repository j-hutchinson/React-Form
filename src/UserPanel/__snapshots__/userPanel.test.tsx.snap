// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPanel component component matches snapshot 1`] = `
<styled.form
  onSubmit={[Function]}
>
  <styled.label>
    <div>
       Name: 
      <styled.div>
        *
      </styled.div>
       
    </div>
    <styled.input
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
  </styled.label>
  <styled.label>
    Role:
    <styled.input
      onChange={[Function]}
      type="text"
      value=""
    />
  </styled.label>
  <styled.label>
    <div>
       Email: 
      <styled.div>
        *
      </styled.div>
       
    </div>
    <styled.input
      onChange={[Function]}
      required={true}
      type="email"
      value=""
    />
  </styled.label>
  <styled.label>
    <div>
       Password: 
      <styled.div>
        *
      </styled.div>
       
    </div>
    <styled.input
      onChange={[Function]}
      required={true}
      type="password"
      value=""
    />
  </styled.label>
  <styled.div>
    <styled.input
      type="submit"
      value="Submit"
    />
  </styled.div>
</styled.form>
`;
